### ISSUE TITLE
[SUGGESTION] Documentation for transactions in v5

### ISSUE THREAD

**Issue by tjmoore** [litedb-org/LiteDB#1881]
I cannot find documentation on the use of transactions in LiteDB 5. The documentation site at https://www.litedb.org/docs/ lacks any talk of transactions as far as I can tell.

I'd like some description on what steps are required and what happens with a 'using' scope and disposal.

e.g. with the following, do I need to call Commit, or is that implicit in disposal? If an exception is thrown is the transaction abandoned? When/why would I call rollback, especially if it's Write-Ahead Logging?

```cs
using (var db = new LiteDatabase("mydb.db"))
{
   db.BeginTrans();

   // do things

   db.Commit();
}
```

Also I can't see much on checkpoint. In typical use, do I need to do anything about checkpoints or just ignore. What happens with checkpointing with exceptions that go out of scope and LiteDatabase is disposed? I assume if there's a complete crash or power failure there's recovery.

I'm trying to piece together bits of information and most I can find is mainly in Gitter chats, and a lot of that dates back to v3.

**Comment by tjmoore** [comment]
#1730

**Comment by tjmoore** [comment]
To add, I'm looking at using TransactionScope to wrap around LiteDatabase and I'm trying to understand the bahaviour to do that.

**Comment by KC7465128305** [comment]
Yes it would be nice if there is some documentation on transaction.

**Comment by 7702244** [comment]
This code works fine for me.

```
using (var db = new LiteDatabase("mydb.db"))
{
   db.BeginTrans();
   try
   {
       // do things
       db.Commit();
   }
   catch (Exception)
   {
       db.Connection.Rollback();
       throw;
   }
}
```

**Comment by tjmoore** [comment]
What I'm not sure about is the `Rollback()` call. If the commit doesn't happen due to exception and given it's Write-Ahead Logging, there's nothing to rollback surely? Or is it a tidy up on an abandoned transaction? More understandable if you keep a LiteDatabase instance which supposedly is more efficient, instead of a using around each transaction block. Then there's no disposal to clean up until you free the instance at a later time or when application shuts down.

As I understand it with a traditional SQL database a rollback journal is used, writes are made live and Rollback on error will undo the transaction using the journal. With Write-Ahead Logging, the transaction is written to a WAL and nothing is changed until Commit, so there's no need for rollback on error?

**Comment by 7702244** [comment]
`Commit()` is necessary. Imagine that before it, there are several operations in a row that depend on each other. What if the last operation throws an exception? We need to roll back all the previous ones.

```
using (var db = new LiteDatabase("mydb.db"))
{
   db.BeginTrans();
   try
   {
       // Replace data with new students.
       db.DeleteAllStudents();
       db.AddNewStudents(); // ---> throws exception, need rollback delete operation
       db.Commit();
   }
   catch (Exception)
   {
       db.Rollback();
       throw;
   }
}
```

**Comment by tjmoore** [comment]
But that's the point of a transaction. The changes aren't committed until Commit. If an exception occurs the transactions aren't applied to the database. As I say, LiteDB 5 is Write-Ahead Logging, like SQLite which as I understand it writes transactions to a separate file and then syncs those changes live on Commit. Unlike traditional rollback journal databases where they are written live and the journal can be used to undo those changes on rollback (transaction recording the start of those changes).

**Comment by Kowjako** [comment]
Same question, how to manage transactions? I saw somewhere that even in single LiteDatabase context, if i will operate on different collections it will treat it as a single transaction?
